Fixes malformed JUnit XML report when running multithreaded

nachomdo opened the pull request:
This PR is intended to fix https://github.com/weavejester/eftest/issues/47

If I understood the problem correctly it seems that `report :begin-test-ns`, `report :end-test-ns` and `report :end-test-var` need to be executed atomically in order to generate a wellformed XML. These changes just delay the computation of `report :begin-test-ns` and `report :end-test-var` by composing the calls to those functions and apply them together when `report :end-test-ns` is called. 

That way the `<testsuite />` element can be written as a whole. I have tested these changes with `:multithread?` true, :vars and :namespaces and seems to generate the XML correctly in all cases.

weavejester commented:
Thanks for the fix! I'll review it, but can you remove commit fece07b? Updating the version number shouldn't be done in PRs.

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
-    (start-suite (name (ns-name (:ns m))))))
+  (let [ns (name (ns-name (:ns m)))
+        f  #(test/with-test-out
+              (start-suite ns))]
```
This could be on one line.

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
-    (finish-suite)))
+  (let [ns (name (ns-name (:ns m)))
+        f  #(test/with-test-out
+              (finish-suite))
```
This could be on one line.

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
+  (let [ns (name (ns-name (:ns m)))
+        f  #(test/with-test-out
+              (finish-suite))
+        g  (get-in @*context* [::deferred-report ns])]
```
It might make more sense for `f` and `g` to be reversed, since `g` is executed before `f`.

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
+        f  #(test/with-test-out
+              (finish-suite))
+        g  (get-in @*context* [::deferred-report ns])]
+    ((juxt g f))
```
This could be `(g) (f)` instead, since we don't care about the output.

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
-          (error-el result)))
-      (finish-case)
-      (swap! *context* update ::test-results dissoc test-var))))
+  (let [ns (-> report/*testing-path*
```
Shouldn't this be:

```clojure
(-> test-var meta :ns ns-name name)
```

?

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
+              (let [test-var (:var m)
+                    results  (get-in @*context* [::test-results test-var])]
+                (start-case (test-name testing-vars)
+                            (name (ns-name (:ns (meta test-var)))))
```
This could be replaced by `ns`, e.g.

```clojure
(start-case (test-name testing-vars) ns)
```

weavejester commented on eftest/src/eftest/report/junit.clj:
```diff
@@ -93,25 +94,37 @@
+                    (error-el result)))
+                (finish-case)
+                (swap! *context* update ::test-results dissoc test-var)))]
+    (swap! *context* update-in [::deferred-report ns] juxt f)))
```
Rather than use `juxt`, which has output, instead:

```clojure
(swap! *context* update-in [::deferred-report ns] (fn [g] (g) (f)))
```

nachomdo commented:
Thanks so much for reviewing this so quickly @weavejester! I think I have addressed all your comments but please let me know if you find anything else that needs to be changed.

weavejester commented:
Tested and it all seems fine, so I'll merge it in. Thanks for the fix!
